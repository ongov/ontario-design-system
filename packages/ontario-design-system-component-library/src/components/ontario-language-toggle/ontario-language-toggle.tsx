import { Component, Prop, State, Event, EventEmitter, Watch, h, Fragment } from '@stencil/core';

import { supportedLanguages, Language } from '../../utils/common/language-types';

import { default as translations } from '../../translations/global.i18n.json';
import { HeaderLanguageToggleEventDetails } from '../../utils/events/common-events.interface';
import { validateValueAgainstArray } from '../../utils/validation/validation-functions';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { printArray, getRootHTMLElement } from '../../utils/helper/utils';

@Component({
	tag: 'ontario-language-toggle',
	styleUrl: 'ontario-language-toggle.scss',
	shadow: true,
})
export class OntarioLanguageToggle {
	/**
	 * The language of the component.
	 *
	 * In most cases, the language toggle should be the source of truth for determining the site language.
	 *
	 * Only pass a language value here if necessary.
	 */
	@Prop({ mutable: true }) language?: Language;

	/**
	 * The size of the language toggle button.
	 *
	 * If no prop is passed, it will be set to the `default` size.
	 */
	@Prop() size?: 'default' | 'small' = 'default';

	/**
	 * The URL to change to when the language toggle button is clicked.
	 *
	 * This is optional.
	 */
	@Prop({ mutable: true }) url?: string;

	/**
	 * A custom function to pass to the language toggle button.
	 *
	 * This is optional.
	 */
	@Prop() customLanguageToggle?: (event: globalThis.Event) => void;

	@State() translations: any = translations;

	/**
	 * Internal state used as the source of truth for component language.
	 */
	@State() private languageState: Language;

	/**
	 * Internal state used to render the text on the language toggle UI.
	 */
	@State() private oppositeLanguageLabel: { fullWord: string; abbreviation: Language } | undefined = undefined;

	/**
	 * Updates the language and languageState props when changes to the language prop are detected.
	 */
	@Watch('language')
	updateLanguage() {
		if (this.language) {
			if (!validateValueAgainstArray(this.language, supportedLanguages)) {
				this.showLanguageWarning(this.language);
				this.language = this.translations.siteLanguage.abbreviation.en as Language;
			}
			this.languageState = this.language;
		}
		this.setAppLanguageHandler();
	}

	/**
	 * Event that fires during the setAppLanguageHandler() method.
	 *
	 * The event contains the current language (after language logic has already occurred).
	 */
	@Event() setAppLanguage: EventEmitter<Language>;

	/**
	 * Event that fires when the language toggle is pressed/clicked.
	 *
	 * The event contains the oldLanguage along with the newLanguage.
	 */
	@Event() headerLanguageToggled: EventEmitter<HeaderLanguageToggleEventDetails>;

	/**
	 * This function sets the languageState (if not already set).
	 *
	 * It also emits the setAppLanguage() event, updates the component language label, and
	 * updates the <html> tag lang attribute with the languageState value.
	 *
	 * It gets called by the connectedCallback() component lifecycle hook, and by the
	 * updateLanguage() method, which is fired on the watch for the language prop.
	 */
	private setAppLanguageHandler() {
		const defaultLang = this.translations.siteLanguage.abbreviation.en;
		const rootLang = getRootHTMLElement()?.lang;

		// If languageState is not set, set it equal to the following cadence:
		// language prop value, <html> tag lang attribute, or default to "en"
		if (!this.languageState) {
			if (this.language) {
				this.languageState = this.language;
			} else if (rootLang) {
				if (validateValueAgainstArray(rootLang, supportedLanguages)) {
					this.languageState = rootLang as Language;
				} else {
					this.showLanguageWarning(rootLang, 'document');
					this.languageState = defaultLang;
				}
			} else {
				this.languageState = defaultLang;
			}
		}

		this.setAppLanguage.emit(this.languageState);

		this.oppositeLanguageLabel = {
			fullWord: this.getOppositeLanguageFullWord(),
			abbreviation: this.getOppositeLanguageAbbrievation(),
		};

		this.updateHTMLLang();
	}

	/**
	 * An event that emits to other components that the language toggle button has been toggled.
	 *
	 * @param {Language} oldLanguage - The language prior to the language toggle being pressed.
	 * @param {globalThis.Event} event - event that triggered the function (e.g. onclick).
	 */
	private handleHeaderLanguageToggled(oldLanguage: Language, event?: globalThis.Event) {
		this.languageState =
			oldLanguage === this.translations.siteLanguage.abbreviation.en
				? this.translations.siteLanguage.abbreviation.fr
				: this.translations.siteLanguage.abbreviation.en;

		this.headerLanguageToggled.emit({ oldLanguage: oldLanguage, newLanguage: this.languageState });

		this.updateHTMLLang();

		if (this.customLanguageToggle && event) {
			this.customLanguageToggle(event);
		}
	}

	/**
	 * Prints a warning message to the console about using an incorrect language for the component.
	 *
	 * @param {string} lang - The incorrect language that was received.
	 * @param {string} type - prop/document | Where the incorrect language is coming from.
	 */
	private showLanguageWarning(lang: string, type: 'prop' | 'document' = 'prop') {
		const propOrDocumentMessage =
			type === 'prop' ? `The language prop value of ${lang} ` : `The HTML document lang attribute value of ${lang} `;
		const message = new ConsoleMessageClass();
		message
			.addDesignSystemTag()
			.addRegularText(propOrDocumentMessage)
			.addRegularText('is not a valid language value for the ')
			.addMonospaceText(' <ontario-language-toggle> ')
			.addRegularText(`component. Valid language values are ${printArray([...supportedLanguages])}. `)
			.addRegularText(`A default language value of ${this.translations.siteLanguage.abbreviation.en} will be applied.`)
			.printMessage();
	}

	/**
	 * Returns abbreviated text for the opposite language.
	 *
	 * @returns {Language}
	 */
	private getOppositeLanguageAbbrievation(): Language {
		return this.languageState === this.translations.siteLanguage.abbreviation.en
			? this.translations.siteLanguage.abbreviation.fr
			: this.translations.siteLanguage.abbreviation.en;
	}

	/**
	 * Returns full word text for the opposite language.
	 *
	 * @returns {string}
	 */
	private getOppositeLanguageFullWord(): string {
		return this.languageState === this.translations.siteLanguage.abbreviation.en
			? this.translations.siteLanguage.fullWord.fr
			: this.translations.siteLanguage.fullWord.en;
	}

	/*
	 * Updates the <html> lang attribute based on component languageState.
	 */
	private updateHTMLLang = () => {
		const htmlElement = getRootHTMLElement();

		if (htmlElement) {
			htmlElement.setAttribute('lang', this.languageState);
		}
	};

	/**
	 * Component life cycle hook.
	 *
	 * https://stenciljs.com/docs/component-lifecycle#connectedcallback
	 */
	connectedCallback() {
		this.setAppLanguageHandler();
	}

	/**
	 * Component life cycle hook.
	 *
	 * https://stenciljs.com/docs/component-lifecycle#componentdidload
	 */
	componentDidLoad() {
		/**
		 * Creates a MutationObserver (a type of watch) on the <html> tag lang attribute.
		 *
		 * When changes occur, the oppositeLanguageLabel state variable regenerates.
		 *
		 * This is to act as a form of callback and create a subtle delay between page content
		 * updating and the language toggle label updating.
		 */
		const observer = new MutationObserver((mutations) => {
			mutations.forEach((mutation) => {
				switch (mutation.type) {
					case 'attributes':
						switch (mutation.attributeName) {
							case 'lang':
								this.oppositeLanguageLabel = {
									fullWord: this.getOppositeLanguageFullWord(),
									abbreviation: this.getOppositeLanguageAbbrievation(),
								};
								break;
						}
						break;
				}
			});
		});

		// Only create/trigger the MutationObserver if the <html> element exists.
		if (getRootHTMLElement()) {
			const options = { attributes: true };
			observer.observe(getRootHTMLElement(), options);
		}
	}

	render() {
		return (
			<a
				aria-label={this.translations.languageToggle.ariaLabel[`${this.oppositeLanguageLabel?.abbreviation}`]}
				class={
					this.size === 'default'
						? 'ontario-language-toggler ontario-language-toggler--default'
						: 'ontario-language-toggler ontario-language-toggler--small'
				}
				href={this.url ? this.url : '#'}
				hreflang={this.oppositeLanguageLabel?.abbreviation}
				lang={this.oppositeLanguageLabel?.abbreviation}
				onClick={(e) => this.handleHeaderLanguageToggled(this.languageState, e)}
			>
				{this.size === 'small' ? (
					<span>{this.oppositeLanguageLabel?.fullWord}</span>
				) : (
					<Fragment>
						<abbr title={this.oppositeLanguageLabel?.fullWord} class="ontario-show-for-small-only">
							{this.oppositeLanguageLabel?.abbreviation.toUpperCase()}
						</abbr>
						<span class="ontario-show-for-medium">{this.oppositeLanguageLabel?.fullWord}</span>
					</Fragment>
				)}
			</a>
		);
	}
}
