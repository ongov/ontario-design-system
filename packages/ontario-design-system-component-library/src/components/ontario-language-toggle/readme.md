# ontario-language-toggle

This component is used internally to toggle the component languages, set translations and update the HTML lang attribute in the DOM.

It renders an anchor tag (styles as an ontario-button) that accepts either a URL prop to switch the language to, or a custom toggle function if additional functionality is necessary.

It is used in the ontario-header component.

<!-- Auto Generated Below -->

## Properties

| Property               | Attribute  | Description                                                                                                                                                                      | Type                                    | Default     |
| ---------------------- | ---------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------- | ----------- |
| `customLanguageToggle` | --         | A custom function to pass to the language toggle button. This is optional.                                                                                                       | `((event: Event) => void) \| undefined` | `undefined` |
| `language`             | `language` | The language of the component. In most cases, the language toggle should be the source of truth for determining the site language. Only pass a language value here if necessary. | `"en" \| "fr" \| undefined`             | `undefined` |
| `size`                 | `size`     | The size of the language toggle button. If no prop is passed, it will be set to the `default` size.                                                                              | `"default" \| "small" \| undefined`     | `'default'` |
| `url`                  | `url`      | The URL to change to when the language toggle button is clicked. This is optional.                                                                                               | `string \| undefined`                   | `undefined` |

## Events

| Event                   | Description                                                                                                                                      | Type                                            |
| ----------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------ | ----------------------------------------------- |
| `headerLanguageToggled` | Event that fires when the language toggle is pressed/clicked. The event contains the oldLanguage along with the newLanguage.                     | `CustomEvent<HeaderLanguageToggleEventDetails>` |
| `setAppLanguage`        | Event that fires during the setAppLanguageHandler() method. The event contains the current language (after language logic has already occurred). | `CustomEvent<"en" \| "fr">`                     |

## Dependencies

### Used by

- [ontario-header](../ontario-header)

### Graph

```mermaid
graph TD;
  ontario-header --> ontario-language-toggle
  style ontario-language-toggle fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
