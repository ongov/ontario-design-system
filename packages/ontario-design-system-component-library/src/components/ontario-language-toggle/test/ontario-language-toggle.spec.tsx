import { newSpecPage } from '@stencil/core/testing';
import { OntarioLanguageToggle } from '../ontario-language-toggle';
import { mutationObserverMock } from '../../../utils/tests/mutation-observer.mock';

global.MutationObserver = mutationObserverMock;

describe('ontario-language-toggle', () => {
	describe('snapshot', () => {
		it('should render the expected html', async () => {
			const page = await newSpecPage({
				components: [OntarioLanguageToggle],
				html: `<ontario-language-toggle></ontario-language-toggle>`,
			});

			expect(page.root).toMatchSnapshot();
		});
	});

	it('should render the default size language toggle if no size is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioLanguageToggle],
			html: `<ontario-language-toggle></ontario-language-toggle>`,
		});

		expect(page.root).toEqualHtml(`
			<ontario-language-toggle>
				<mock:shadow-root>
					<a aria-label="Cliquez pour changer la langue en français" class="ontario-language-toggler ontario-language-toggler--default" href="#" hreflang="fr" lang="fr">
						<abbr class="ontario-show-for-small-only" title="Français">
							FR
						</abbr>
						<span class="ontario-show-for-medium">
							Français
						</span>
					</a>
				</mock:shadow-root>
			</ontario-language-toggle>
		`);
		expect(page.rootInstance.size).toBe('default');
	});

	/**
	 * TODO: Determine if this is testable.
	 *
	 * Not sure how to properly test against this, seems to be a race condition.
	 *
	 * Takes a few processes before the html lang attribute is updated.
	 *
	 * Using setTimeout() leads to a false positive.
	 *
	 * Using page.waitForChanges() or autoApplyChanges: true seem to have no effect.
	 */
	it.skip('should render a default language of English on the <html> tag if no language prop is passed', async () => {
		const page = await newSpecPage({
			components: [OntarioLanguageToggle],
			html: `<ontario-language-toggle></ontario-language-toggle>`,
		});

		expect(page.rootInstance.language).toBe('en');
	});

	it('should render a small size language toggle with a French default language when explicitly specified', async () => {
		const page = await newSpecPage({
			components: [OntarioLanguageToggle],
			html: `<ontario-language-toggle size="small" url="/en" language="fr"></ontario-language-toggle>`,
		});

		expect(page.root).toEqualHtml(`
			<ontario-language-toggle size="small" url="/en" language="fr">
				<mock:shadow-root>
					<a aria-label="Click to switch the language to English" class="ontario-language-toggler ontario-language-toggler--small" href="/en" hreflang="en" lang="en">
						<span>
							English
						</span>
					</a>
				</mock:shadow-root>
			</ontario-language-toggle>
		`);
	});
});
