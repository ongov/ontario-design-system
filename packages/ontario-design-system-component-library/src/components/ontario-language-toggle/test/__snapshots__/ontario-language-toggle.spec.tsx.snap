// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-language-toggle snapshot should render the expected html 1`] = `
<ontario-language-toggle>
  <template shadowrootmode="open">
    <a aria-label="Cliquez pour changer la langue en français" class="ontario-language-toggler ontario-language-toggler--default" href="#" hreflang="fr" lang="fr">
      <abbr class="ontario-show-for-small-only" title="Français">
        FR
      </abbr>
      <span class="ontario-show-for-medium">
        Français
      </span>
    </a>
  </template>
</ontario-language-toggle>
`;
